recursive struct with array

The following program `sample.go` triggers a panic:

```go
package main

import "fmt"

type AST struct {
    Num      int
    Children []AST
}

func newAST(num int, root AST, children ...AST) AST {
    return AST{num, append([]AST{root}, children...)}
}

func main() {
    ast := newAST(1, AST{})
    fmt.Println(ast)
}
```

Expected result:
```console
$ go run ./sample.go
{1 [{0 []}]}
```

Got:
```console
$ yaegi ./sample.go
/Users/marc/sample.go:11:18: panic
/Users/marc/sample.go:15:9: panic
run: reflect.Set: value of type []struct { Num int; Children []interface {} } is not assignable to type struct { Num int; Children []interface {} }
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0x14000190e00, 0x40)
	/Users/marc/sdk/go1.16.2/src/runtime/debug/stack.go:24 +0x88
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0x140004a3c40)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:491 +0x94
panic(0x102ddd8e0, 0x140001bb170)
	/Users/marc/sdk/go1.16.2/src/runtime/panic.go:965 +0x14c
github.com/traefik/yaegi/interp.runCfg.func1(0x1400015cc60, 0x1400048fe00, 0x140004a3a10)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:185 +0x1e8
panic(0x102ddd8e0, 0x140001bb170)
	/Users/marc/sdk/go1.16.2/src/runtime/panic.go:965 +0x14c
github.com/traefik/yaegi/interp.runCfg.func1(0x1400015cdc0, 0x1400048ee00, 0x140004a37f0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:185 +0x1e8
panic(0x102ddd8e0, 0x140001bb170)
	/Users/marc/sdk/go1.16.2/src/runtime/panic.go:965 +0x14c
reflect.Value.assignTo(0x14000300840, 0x1400012acf0, 0x197, 0x102bffba4, 0xb, 0x1400013f080, 0x0, 0x97, 0x1400012ad68, 0x14000300840)
	/Users/marc/sdk/go1.16.2/src/reflect/value.go:2451 +0x2d0
reflect.Value.Set(0x1400013f080, 0x140003009e0, 0x199, 0x14000300840, 0x1400012acf0, 0x197)
	/Users/marc/sdk/go1.16.2/src/reflect/value.go:1564 +0x94
reflect.Append(0x14000300840, 0x1400012ad20, 0x197, 0x140004a37a0, 0x1, 0x1, 0x1400012ad50, 0x97, 0x199)
	/Users/marc/sdk/go1.16.2/src/reflect/value.go:2051 +0xac
github.com/traefik/yaegi/interp._append.func2(0x1400015cdc0, 0x140004318c0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:3008 +0xf0
github.com/traefik/yaegi/interp.runCfg(0x1400048ee00, 0x1400015cdc0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:191 +0x8c
github.com/traefik/yaegi/interp.call.func7(0x1400015cc60, 0x140004a62a0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:1305 +0x998
github.com/traefik/yaegi/interp.runCfg(0x1400048fe00, 0x1400015cc60)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:191 +0x8c
github.com/traefik/yaegi/interp.(*Interpreter).run(0x14000372000, 0x1400048f300, 0x1400015cb00)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:122 +0x244
github.com/traefik/yaegi/interp.(*Interpreter).eval(0x14000372000, 0x1400010c900, 0xf6, 0x16da3b8e0, 0x1e, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:583 +0x5e8
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0x14000372000, 0x16da3b8e0, 0x1e, 0x1400010c800, 0xf6, 0x0, 0x0, 0x102bed690)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:405 +0xc4
main.runFile(0x14000372000, 0x16da3b8e0, 0x1e, 0x1, 0x1)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:123 +0x84
main.run(0x14000140010, 0x1, 0x1, 0x14000102480, 0x1400006fea8)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:89 +0x724
main.main()
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x324
```

--- a/_test/issue-1065.go
+++ b/_test/issue-1065.go
@@ -1,0 +1,20 @@
+package main
+
+import "fmt"
+
+type AST struct {
+	Num      int
+	Children []AST
+}
+
+func newAST(num int, root AST, children ...AST) AST {
+	return AST{num, append([]AST{root}, children...)}
+}
+
+func main() {
+	ast := newAST(1, AST{}, AST{})
+	fmt.Println(ast)
+}
+
+// Output:
+// {1 [{0 []} {0 []}]}

--- a/interp/run.go
+++ b/interp/run.go
@@ -2947,7 +2947,7 @@ func _append(n *node) {
 	if len(n.child) == 3 {
 		c1, c2 := n.child[1], n.child[2]
 		if (c1.typ.cat == valueT || c2.typ.cat == valueT) && c1.typ.rtype == c2.typ.rtype ||
-			c2.typ.cat == arrayT && c2.typ.val.id() == n.typ.val.id() ||
+			(c2.typ.cat == arrayT || c2.typ.cat == variadicT) && c2.typ.val.id() == n.typ.val.id() ||
 			isByteArray(c1.typ.TypeOf()) && isString(c2.typ.TypeOf()) {
 			appendSlice(n)
 			return

--- a/interp/typecheck.go
+++ b/interp/typecheck.go
@@ -48,6 +48,10 @@ func (check typecheck) assignment(n *node, typ *itype, context string) error {
 		return nil
 	}
 
+	if typ.isRecursive() || typ.val != nil && typ.val.isRecursive() {
+		return nil
+	}
+
 	if !n.typ.assignableTo(typ) {
 		if context == "" {
 			return n.cfgErrorf("cannot use type %s as type %s", n.typ.id(), typ.id())

